pyop2/types/glob.py

```diff
@@ -15,48 +16,28 @@
 from pyop2.types.data_carrier import DataCarrier, EmptyDataMixin, VecAccessMixin
 
 
-class Global(DataCarrier, EmptyDataMixin, VecAccessMixin):
-
-    """OP2 global value.
-
-    When a ``Global`` is passed to a :func:`pyop2.op2.par_loop`, the access
-    descriptor is passed by `calling` the ``Global``.  For example, if
-    a ``Global`` named ``G`` is to be accessed for reading, this is
-    accomplished by::
-
-      G(pyop2.READ)
-
-    It is permissible to pass `None` as the `data` argument.  In this
-    case, allocation of the data buffer is postponed until it is
-    accessed.
-
-    .. note::
-        If the data buffer is not passed in, it is implicitly
-        initialised to be zero.
-    """
-
-    _modes = [Access.READ, Access.INC, Access.MIN, Access.MAX]
+class SetFreeDataCarrier(DataCarrier, EmptyDataMixin):
 
     @utils.validate_type(('name', str, ex.NameTypeError))
-    def __init__(self, dim, data=None, dtype=None, name=None, comm=None):
-        if isinstance(dim, Global):
-            # If g is a Global, Global(g) performs a deep copy. This is for compatibility with Dat.
+    def __init__(self, dim, data=None, dtype=None, name=None):
+        if issubclass(type(dim), Global):
+            # If g is a Global, Global(g) performs a deep copy.
+            # This is for compatibility with Dat.
             self.__init__(dim._dim, None, dtype=dim.dtype,
                           name="copy_of_%s" % dim.name, comm=dim.comm)
             dim.copy(self)
+        elif issubclass(type(dim), Constant):
+            # If g is a Local, Local(g) performs a deep copy.
+            # This is for compatibility with Dat.
+            self.__init__(dim._dim, None, dtype=dim.dtype,
+                          name="copy_of_%s" % dim.name)
+            dim.copy(self)
```

Comment:
These should be moved into the constructors of the subclasses. It's very odd having this here. Also note the error:
> `If g is a Local`

(should be `Constant`)